crates/core/tedge_monitor/src/main.rs

```diff
@@ -0,0 +1,150 @@
+use futures::stream::FuturesUnordered;
+use futures::StreamExt;
+use mqtt_channel::{Config, Message, MqttError, PubChannel, Topic};
+use nanoid::nanoid;
+use std::process::{self, Command, Stdio};
+
+use serde::{Deserialize, Serialize};
+use tedge_config::{
+    ConfigRepository, ConfigSettingAccessor, ConfigSettingError, MqttBindAddressSetting,
+    MqttPortSetting, TEdgeConfigError,
+};
+use tokio::time::Duration;
+
+pub const DEFAULT_TEDGE_CONFIG_PATH: &str = "/etc/tedge";
+
+#[derive(Serialize, Deserialize)]
+pub struct Response {
+    status: String,
+    pid: u32,
+}
+#[tokio::main]
+async fn main() {
+    let _ = start_watchdog().await;
+}
+
+async fn start_watchdog() {
+    // Send ready notification to systemd.
+    let _ = notify_systemd(process::id(), "--ready");
+
+    // Start helth check request publisher
+    tokio::spawn(async move { publish().await });
+
+    loop {
+        let tedge_services = vec![
+            "tedge-mapper-c8y",
+            "tedge-mapper-az",
+            "tedge-mapper-collectd",
+            "tedge-agent",
+        ];
+        let watchdog_threads = FuturesUnordered::new();
+
+        for service in tedge_services {
+            let res_topic = format!("tedge/health/{}", service);
+
+            watchdog_threads.push(tokio::spawn(async move {
+                monitor_tedge_service(service, &res_topic).await.unwrap()
+            }));
+        }
+
+        futures::future::join_all(watchdog_threads).await;
+    }
+}
+
+async fn publish() -> Result<(), WatchdogError> {
+    let client_id: &str = "watchdog_publisher";
+    let mqtt_config = get_mqtt_config(client_id)?;
+
+    let client = mqtt_channel::Connection::new(&mqtt_config).await?;
+
+    let mut publisher = client.published;
+    let topic = Topic::new("tedge/health-check")?;
+
+    loop {
+        let message = Message::new(&topic, &b"\0"[..]);
+        match publisher.publish(message).await {
+            Ok(()) => {}
+            Err(e) => {
+                eprintln!("Publish failed with error: {}", e.to_string());
+            }
+        }
+        tokio::time::sleep(Duration::from_secs(2)).await;
+    }
+}
+
+async fn monitor_tedge_service(name: &str, res_topic: &str) -> Result<(), WatchdogError> {
+    let client_id: &str = &format!("{}_{}", name, nanoid!());
+    let mqtt_config = get_mqtt_config(client_id)?.with_subscriptions(res_topic.try_into()?);
+    let client = mqtt_channel::Connection::new(&mqtt_config).await?;
```

Comment:
This means that we establish a new MQTT connection on every loop iteration , right? I think we should just connect once and then just do the timed response processing in every loop.

Reply:
Refactored